File: button.cpp
#include "button.h"

Button::Button(size_t count) : pressed_(count, false) {}

const char* Button::Name() const {
    return "Button";
}

void Button::Reset() {
    for (size_t i = 0; i < pressed_.size(); ++i) {
        pressed_[i] = false;
    }
}

void Button::Tick() {
    // Buton bu örnekte pasif.
}

void Button::SetPressed(size_t index, bool pressed) {
    if (index < pressed_.size()) {
        pressed_[index] = pressed;
    }
}

bool Button::IsPressed(size_t index) const {
    if (index < pressed_.size()) {
        return pressed_[index];
    }
    return false;
}

size_t Button::Count() const {
    return pressed_.size();
}

File: button/button.go
package button

// Button basılma durumunu memory map üzerinden okunan birim olarak sunar.
type Button struct {
	name    string
	addr    uint32
	pressed bool
}

func New(name string, addr uint32) *Button { return &Button{name: name, addr: addr} }
func (b *Button) Name() string             { return b.name }
func (b *Button) Reset()                   { b.pressed = false }
func (b *Button) Tick(_ uint64)            {}

func (b *Button) SetPressed(v bool) { b.pressed = v }

func (b *Button) Read(addr uint32) (uint32, bool) {
	if addr != b.addr {
		return 0, false
	}
	if b.pressed {
		return 1, true
	}
	return 0, true
}

func (b *Button) Write(addr uint32, value uint32) bool {
	if addr != b.addr {
		return false
	}
	b.pressed = value&1 == 1
	return true
}
